pkg/controller/uniregistrysyncer/uniregistrysyncer.go

```diff
@@ -0,0 +1,275 @@
+		if !filter(in.NamespacedName) {
+			return nil
+		}
+		return []reconcile.Request{in}
+	})
+}
+
+type reconciler struct {
+	log         *logrus.Entry
+	appCIClient ctrlruntimeclient.Client
+	apiCIClient ctrlruntimeclient.Client
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(ctx, req, log)
+	// Ignore the logging for IsConflict errors because they are results of concurrent reconciling
+	if err != nil && !apierrors.IsConflict(err) && !apierrors.IsAlreadyExists(err) {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+const (
+	annotationDPTPRequester = "dptp.openshift.io/requester"
+)
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, log *logrus.Entry) error {
+	sourceImageStreamTag := &imagev1.ImageStreamTag{}
+	sourceClient := r.appCIClient
+	if err := sourceClient.Get(ctx, req.NamespacedName, sourceImageStreamTag); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("the source imageStreamTag not found")
+			return nil
+		}
+		return fmt.Errorf("failed to get source imageStreamTag %s: %w", req.NamespacedName.String(), err)
+	}
+
+	imageStreamNameAndTag := strings.Split(req.Name, ":")
+	if n := len(imageStreamNameAndTag); n != 2 {
+		return fmt.Errorf("when splitting imagestreamtagname %s by : expected two results, got %d", req.Name, n)
+	}
+	imageStreamName, imageTag := imageStreamNameAndTag[0], imageStreamNameAndTag[1]
+	isName := types.NamespacedName{Namespace: req.Namespace, Name: imageStreamName}
+	sourceImageStream := &imagev1.ImageStream{}
+	if err := sourceClient.Get(ctx, isName, sourceImageStream); err != nil {
+		// received a request on the isTag, but the 'is' is no longer there
+		return fmt.Errorf("failed to get source imageStream %s: %w", isName.String(), err)
+	}
+
+	*log = *log.WithField("docker_image_reference", sourceImageStreamTag.Image.DockerImageReference)
+
+	targetClient := r.apiCIClient
+	isCurrent, err := r.isImageStreamTagCurrent(ctx, req.NamespacedName, targetClient, sourceImageStreamTag)
+	if err != nil {
+		return fmt.Errorf("failed to check if imageStreamTag on the target cluster is current: %w", err)
+	}
+	if isCurrent {
+		log.Debug("ImageStreamTag is current")
+		return nil
+	}
+
+	if err := targetClient.Get(ctx, types.NamespacedName{Name: req.Namespace}, &corev1.Namespace{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check if namespace %s exists on the target cluster: %w", req.Namespace, err)
+		}
+		if err := targetClient.Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{
+			Name:        req.Namespace,
+			Annotations: map[string]string{annotationDPTPRequester: ControllerName},
```

Comment:
I think we have different annotations for this elsewhere, can we check on that and try to use the same annotation everywhere?

Reply:
Found other places: They use label instead of annotations. I am going to use label here too.
```
Labels:    map[string]string{"ci.openshift.org/auto-managed": "true"},
Labels["ci.openshift.org/secret-syncer-controller-managed"] = "true"
```

Reply:
Changed my mind.
The 2 controllers use the same constant from `api` pkg.

I think we should change the other tools to use annotations too.

Reply:
> I think we should change the other tools to use annotations too.

Why? Labels allow filtering, annotations do not. What reason do you see to prefer Annotations over Labels?

Reply:
I was thinking openshift does ("openshift.io/requester": "hongkailiu" as an annotation) on namespace. I guess we should do the same?


```bash
$ oc get ns ci-op-223-test -o  yaml | yq '.metadata.annotations'
{
  "openshift.io/description": "",
  "openshift.io/display-name": "",
  "openshift.io/requester": "hongkailiu",
  ...
}
```

Reply:
Updated the controllers.
Can I do those 2 with another PR?

```
Labels:    map[string]string{"ci.openshift.org/auto-managed": "true"},
Labels["ci.openshift.org/secret-syncer-controller-managed"] = "true"
```